### PR TITLE
Move the `lightning_optimizers` ownership to the `Strategy`

### PR DESCRIPTION
**Pull request by carmocca**
### Context

Currently, the `Strategy` is the object holding the optimizers. We have an internal wrapper called the `LightningOptimizer` which maps 1:1 to each `Optimizer`.

The list of lightning optimizers is currently owned by the `Trainer`. There's no requirement for that anymore so this PR moves the ownership to the `Strategy`, together with the optimizer references.

### Benefits

- Each time the `Strategy.optimizers` list gets set, the `strategy._lightning_optimzers` dictionary des too, this avoids stale references.
- Removes clutter from the `Trainer`
- Allows merging the `LightningOptimizer` and the `_LiteOptimizer` in a follow-up: #11452
- Deprecates `trainer.lightning_optimizers`, as there's no reason why users would need to access or know about this internal wrapper.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @justusschock @awaelchli @akihironitta @rohitgr7 @tchaton

### CODE REVIEW

File: pytorch_lightning/core/optimizer.py [hunk 1]
```diff
@@ -164,17 +153,14 @@ def closure_dis():
             profiler_action = "optimizer_step_with_closure"
         profiler_action += f"_{self._optimizer_idx}"
 
-        trainer = self._trainer
-        assert trainer is not None
-        with trainer.profiler.profile(profiler_action):
-            trainer.strategy.optimizer_step(self._optimizer, self._optimizer_idx, closure, **kwargs)
+        with self._lightning_module.trainer.profiler.profile(profiler_action):
```
Comment (carmocca) [inline pytorch_lightning/core/optimizer.py:156]: This is very ugly but the path of least friction at the moment.

I guess this profiler call will need to be removed. It could only be moved to the loops, but they already profile it through `call_hook`.

Means we will lose the closure suffix.

Left for a follow-up

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -2014,13 +2004,13 @@ def optimizers(self) -> List[Optimizer]:
...
-        # Necessary to rewrap optimizers to lightning
-        # They will be re-created when accessing
-        # the `lightning_optimizers` trainer property
-        self._lightning_optimizers = None
-
         self.strategy.optimizers = new_optims
 
+    @property
+    def lightning_optimizers(self) -> Dict[int, LightningOptimizer]:
+        # FIXME: why do we expose this?
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:2011]: Only two places accesss this property:

- `OptimizerLoop._optimizer.step`
- `LightningModule.optimizers`

So I guess we could deprecate it, yay or nay?

File: pytorch_lightning/strategies/strategy.py [hunk 3]
```diff
@@ -84,6 +85,27 @@ def precision_plugin(self) -> PrecisionPlugin:
...
+    def optimizers(self, optimizers: List[Optimizer]) -> None:
+        self._optimizers = optimizers
+        self._lightning_optimizers.clear()
+
+    @property
+    def _lightning_optimizers(self) -> Dict[int, LightningOptimizer]:
+        if not self._cached_lightning_optimizers:
+            # we create this on-the-fly instead of only in the optimizers setter because deleting this reference
+            # after every batch is supported as indicated by the test:
+            # tests/core/test_lightning_optimizer.py::test_lightning_optimizer_keeps_hooks
```
Comment (carmocca) [inline pytorch_lightning/strategies/strategy.py:102]: I don't think doing this is sensible.

The test in question was added in #https://github.com/PyTorchLightning/pytorch-lightning/pull/6305 but I don't think it intended that deleting the lightning_optimizers per batch is supported.

We can also update the test and move this back to `optimizers.setter`

Up for discussion

File: pytorch_lightning/strategies/strategy.py [hunk 4]
```diff
@@ -51,7 +51,8 @@ def __init__(
         self._model: Optional[Module] = None
         self.checkpoint_io = checkpoint_io
         self.precision_plugin = precision_plugin
-        self.optimizers: List[Optimizer] = []
+        self._optimizers: List[Optimizer] = []
+        self._cached_lightning_optimizers: Dict[int, LightningOptimizer] = {}
```
Comment (carmocca) [inline pytorch_lightning/strategies/strategy.py:55]: We could also not cache this at all and re-generate it every time it's necessary (num training batches * num optimizers).

It would simplify the code, but not sure if it's a good idea.

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -2519,15 +2505,6 @@ def weights_summary(self, val: Optional[str]) -> None:
     Other
     """
 
-    # TODO: refactor this so that it can be done in LightningOptimizer
-    def __getstate__(self):
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:2523]: I'm not sure if we need this in the strategy, letting CI check

File: tests/trainer/test_trainer.py [hunk 6]
```diff
@@ -1763,27 +1763,6 @@ def backward(self, *args, **kwargs):
     ]
 
 
-def test_init_optimizers_resets_lightning_optimizers(tmpdir):
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:1766]: Moved to `tests/core/test_lightning_optimizer.py` with no changes othger than the property access

File: tests/core/test_lightning_optimizer.py [hunk 7]
```diff
@@ -287,25 +287,30 @@ def step(self, closure=None):
...
-        def on_train_batch_start(self, batch: Any, batch_idx: int) -> None:
-            self.count_on_train_batch_start += 1
-            optimizer = self.optimizers(use_pl_optimizer=False)
-            assert len(optimizer._fwd_handles) == 1
+def test_init_optimizers_resets_lightning_optimizers(tmpdir):
+    """Test that the Trainer resets the `lightning_optimizers` list everytime new optimizers get initialized."""
 
-        def on_train_batch_end(self, outputs: Any, batch: Any, batch_idx: int) -> None:
-            self.count_on_train_batch_end += 1
-            del self.trainer._lightning_optimizers
```
Comment (carmocca) [inline tests/core/test_lightning_optimizer.py:304]: I don't think allowing `del lightning_optimizers` during fitting makes sense.

Refactored the test to not use `fit`

File: pytorch_lightning/core/optimizer.py [hunk 8]
```diff
@@ -164,17 +159,16 @@ def closure_dis():
             profiler_action = "optimizer_step_with_closure"
         profiler_action += f"_{self._optimizer_idx}"
 
-        trainer = self._trainer
-        assert trainer is not None
-        with trainer.profiler.profile(profiler_action):
-            trainer.strategy.optimizer_step(self._optimizer, self._optimizer_idx, closure, **kwargs)
+        assert self._strategy is not None
+        assert self._strategy.lightning_module is not None
+        with self._strategy.lightning_module.trainer.profiler.profile(profiler_action):
```
Comment (carmocca) [inline pytorch_lightning/core/optimizer.py:164]: This is very ugly but the path of least friction at the moment.

I guess this profiler call will need to be removed. It could only be moved to the loops, but they already profile it through `call_hook`.

Means we will lose the closure suffix.

Left for a follow-up

**Comment by awaelchli** [comment]
apart from the comments here, there is no issue or description. for others it will be hard to get the picture of what you briefly explained offline. would you mind adding that?

**Comment by carmocca** [comment]
Updated the top post @awaelchli. If you think this PR needs further discussion, I can also open an issue if you prefer that.

File: pytorch_lightning/loops/optimization/optimizer_loop.py [hunk 9]
```diff
@@ -351,8 +351,13 @@ def _optimizer_step(
...
         """
         is_lbfgs = isinstance(optimizer, torch.optim.LBFGS)
+
         # wraps into LightningOptimizer only for running step
-        optimizer = LightningOptimizer._to_lightning_optimizer(optimizer, self.trainer, opt_idx)
+        if self.trainer.amp_backend == AMPType.APEX:
+            # apex overrides .step function and need to be wrapped on each step
+            optimizer = LightningOptimizer._to_lightning_optimizer(optimizer, self.trainer.strategy, opt_idx)
+        else:
+            optimizer = self.trainer.strategy._lightning_optimizers[opt_idx]
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimization/optimizer_loop.py:360]: we will need to think about how to get rid of this later on 👍

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/optimizers.py [hunk 10]
```diff
@@ -52,4 +48,17 @@ def convert_to_lightning_optimizers(self):
...
+            if not isinstance(optimizer, LightningOptimizer):
+                optimizer = LightningOptimizer(optimizer)  # type: ignore [assignment]
+            optimizer._trainer = self
+            for opt_idx, opt in enumerate(self.optimizers):
+                if opt == optimizer._optimizer:
+                    optimizer._optimizer_idx = opt_idx
+                    break
+            return optimizer  # type: ignore [return-value]
+
+        self.strategy._cached_lightning_optimizers = {  # type: ignore [assignment]
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/optimizers.py:62]: In the `strategy._lightning_optimizers` getter

File: pytorch_lightning/core/lightning.py [hunk 11]
```diff
@@ -148,7 +148,7 @@ def optimizers(
             A single optimizer, or a list of optimizers in case multiple ones are present.
         """
         if use_pl_optimizer:
-            opts = list(self.trainer.lightning_optimizers.values())
+            opts = list(self.trainer.strategy._lightning_optimizers.values())
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/core/lightning.py:151]: I believe it was because of this test.

https://github.com/PyTorchLightning/pytorch-lightning/pull/11444#discussion_r783360396

But I don't know why we would want to support destroying the lightning optimizers inside the training step

File: pytorch_lightning/loops/utilities.py [hunk 12]
```diff
@@ -162,19 +162,19 @@ def _update_dataloader_iter(data_fetcher: AbstractDataFetcher, batch_idx: int) -
 
 
 @contextmanager
-def _block_parallel_sync_behavior(trainer: "pl.Trainer", block: bool = True) -> Generator[None, None, None]:
+def _block_parallel_sync_behavior(strategy: Strategy, block: bool = True) -> Generator[None, None, None]:
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/loops/utilities.py:165]: This change is required because LightningOptimizer calls this function but it does not have a trainer reference anymore

File: pytorch_lightning/loops/optimization/optimizer_loop.py (hunk 9 above)
Comment (carmocca) [reply pytorch_lightning/loops/optimization/optimizer_loop.py:360]: Agree

File: pytorch_lightning/core/lightning.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:151]: This is legacy behavior and should be removed. I believe this was meant to support dynamic optimizer, but we never added that feature.

File: pytorch_lightning/core/optimizer.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/core/optimizer.py:164]: This was there to support manual optimization with profiling.

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by justusschock — APPROVED** [review]
